### PR TITLE
feat(messaging): support Google Chat on Hermes over keyless Pub/Sub REST pull

### PR DESCRIPTION
**Pull request by hunglp6d**
## Summary
Google Chat was the only messaging channel restricted to OpenClaw. This change enables it for Hermes without placing the service-account key inside the sandbox: Hermes pulls Chat events from the configured Pub/Sub subscription over the REST API and replies through the OpenShell L7 proxy with a gateway-minted bearer, so the sandbox only ever holds a credential placeholder. Before, `channels add googlechat` was refused on a Hermes sandbox; after, the channel enrolls, receives DMs, and replies.

## Related Issue
Completes the Hermes half of #5492 for Google Chat. That issue models support per agent and states a channel needs to work for each runtime whose upstream actually supports it; Hermes `v2026.7.20` ships `plugins/platforms/google_chat/adapter.py`, so Hermes is inside its accepted scope. #7317 delivered the OpenClaw half and the issue was closed with that half done. It is reopened, and its channel table now records this PR against the Google Chat row. Not a closing reference, since the issue tracks the whole channel catalogue. Release target: next patch release.

## Changes
- Add `channels/googlechat/policy/hermes.yaml`: Pub/Sub REST pull for inbound and Chat REST for the reply, restricted to the two Pub/Sub operations the adapter issues (`:pull`, `:acknowledge`) and to the Chat `spaces` tree for writes.
- Add `channels/googlechat/provider-profile/hermes.yaml`: one gateway-minted token covering `chat.bot` and `pubsub`, with the service-account private key designated as gateway-side secret material.
- Extend `channels/googlechat/manifest.ts`: allow Hermes, add the Hermes-only project and subscription inputs, render the Hermes env and platform fragment, and declare the `google-*` packages the managed image needs.
- Move `google_chat` from the managed-image neutral list to the Hermes supported list, and let the bridge-provider collector select a profile by agent.
- Add `channels/googlechat/runtime/hermes-adapter.py`: a channel-owned runtime asset that subclasses the bundled Google Chat adapter for a Pub/Sub REST pull loop, placeholder credentials, and an aiohttp reply transport, and attaches it through `platform_registry.get()` plus `dataclasses.replace()`.
- Correct the DM allowlist prompt, which told every operator to use `users/NNN` ids and claimed emails are ignored. That holds for OpenClaw and is inverted for Hermes, and the prompt did not say that filling the allowlist switches the DM policy from pairing to allowlist, so a wrong-form entry drops the sender with no reply and no pairing code.

**Why the module load is conditional.** Hermes loads only `__init__.py` as the plugin entry, so the gate in `agents/hermes/plugin/__init__.py` loads the sibling channel asset by path when the rendered `GOOGLE_CHAT_SUBSCRIPTION_NAME` is present. Nothing else reads the module, and the registry entry is replaced through `platform_registry.get()` and `dataclasses.replace()` rather than a global `register` wrapper. A failed load logs and returns without aborting plugin registration. The gate itself has no automated test, for the reason recorded under Quality Gates.

## Type of Change
- [x] Code change (feature, bug fix, or refactor)
- [ ] Code change with doc updates
- [ ] Doc only (prose changes, no code sample modifications)
- [ ] Doc only (includes code sample changes)

## Quality Gates
- [x] Tests added or updated for changed behavior
- [ ] Existing tests cover changed behavior — justification:
- [ ] Tests not applicable — justification:
- [x] Sensitive paths changed (security, policy, credentials, preflight, onboarding, inference, runner, sandbox, or messaging)
- [ ] Sensitive-path review completed or maintainer-approved waiver recorded — reviewer/approval link/justification:
- [ ] Non-success, skipped, or missing CI check accepted by maintainer — check name, approval link, and follow-up issue:

**Coverage gap.** The override itself is covered: `channels/googlechat/runtime/hermes-adapter.test.ts` drives the real pull loop under `python3`, and `channels/googlechat/policy.test.ts` pins the egress preset. What has no automated coverage is the load gate in `agents/hermes/plugin/__init__.py`, because this repository runs no CI lane for Python tests under `agents/hermes/plugin/`, so a unit test placed there would gate nothing. The risk it would guard against is a runtime property, that the gate stops firing and Hermes silently keeps the stock gRPC and service-account adapter, which the REST-only policy then blocks. The nearest real coverage is the `hermes-e2e` lane. Live validation is recorded under Verification.

## DGX Station Hardware Evidence
- [ ] Tested on DGX Station
- Tested commit:
- Station profile/scenario:
- Result:
- Supporting evidence:

## Verification
- [x] PR description includes a `Signed-off-by:` line and every commit appears as `Verified` in GitHub
- [x] Normal `pre-commit`, `commit-msg`, and `pre-push` hooks passed, or `npm run validate:pr` passed after refreshing `origin/main` when hooks were skipped or unavailable
- [x] Targeted behavior tests pass for the current change set, or tests are marked not applicable above — command/result or justification: `npx vitest run src/lib/messaging/channels --project cli --coverage=false` passes 410 tests across 36 files; `npx vitest run src/lib/messaging/applier/setup-applier.test.ts --project cli --coverage=false` passes 24 tests. Both were run after merging `origin/main` into the branch. `src/lib/messaging/channels/googlechat/tunnel/lifecycle.test.ts` needs the compiled plugin, so run `npm --prefix nemoclaw run build` first.
- [ ] Applicable broad gate passed — `npm test` for broad runtime/test-harness changes; `npm run check` for repo-wide validation/coverage changes — command/result:
- [x] Quality Gates section completed with required justifications or waivers
- [x] No secrets, API keys, or credentials committed
- [ ] `npm run docs` builds without warnings (doc changes only)
- [ ] Doc pages follow the [style guide](https://github.com/NVIDIA/NemoClaw/blob/main/docs/CONTRIBUTING.md) (doc changes only)
- [ ] New doc pages include SPDX header and frontmatter (new pages only)

**Live validation.** On a managed Hermes sandbox built from this branch, the plugin replaces the bundled `google_chat` entry, `connect()` reports the keyless REST pull transport, a Chat DM reaches the agent, and the bot replies in the space. The conditional load was exercised against the baked module inside that sandbox in both directions: with the channel configured the module loads and the entry is replaced; without it the module is not loaded and the bundled entry stands.

---
Signed-off-by: Hung Le <hple@nvidia.com>

## Summary by CodeRabbit

- **New Features**
  - Added Google Chat support for Hermes alongside OpenClaw.
  - Added configuration for Google Cloud projects, Pub/Sub subscriptions, service accounts, and message allowlists.
  - Added secure message receiving and acknowledgement through Google Chat and Pub/Sub.
  - Added agent-specific setup and provider configuration for Google Chat integrations.

- **Bug Fixes**
  - Improved resilience during temporary connection and acknowledgement failures, including message retries.
  - Added preflight checks to block setup when required messaging credentials are missing.

- **Tests**
  - Added coverage for setup, permissions, delivery, retries, and configuration resolution.

### CODE REVIEW

**Review by jyaunches — DISMISSED** [review]
## LOC Reduction / Codebase Simplicity Review

### Why this blocks

The PR adds a 477-line repository-local fork of Hermes’s Google Chat behavior. It duplicates `connect()` at `agents/hermes/plugin/googlechat_sandbox_adapter.py:200`, rebinds four private methods at line 392, and wraps and mutates the global platform registry at line 428. A second activation path is hard-coded in `agents/hermes/plugin/__init__.py:1442-1490`.

That is 477 of 953 added lines, creates two owners for the Hermes adapter contract, and depends on private implementation details and deferred registration order. No Python or runtime test covers this layer; the PR text identifies the registry-order behavior as untested.

This also conflicts with `src/lib/messaging/AGENTS.md:8-10` and `src/lib/messaging/AGENTS.md:69-77`, which place channel boot and connect shims under channel-owned manifest/runtime assets with behavior coverage.

### Refactor direction

- Add or use a narrow, tested Hermes extension seam for transport and credentials that preserves the registered adapter metadata.
- Keep only the Pub/Sub REST pull and placeholder-auth delta in NemoClaw, selected through the Google Chat channel runtime.
- Remove the copied `connect()`, global `platform_registry.register` wrapper, private-method rebinding, and separate global-plugin channel gate.
- If the pinned Hermes release lacks a stable seam, land that seam before presenting Hermes Google Chat as supported.

### Expected result

One manifest-owned activation path, deterministic tests, and hundreds fewer lines. NemoClaw would maintain only its REST and credential delta instead of a parallel Hermes adapter implementation.

**Comment by hunglp6d** [comment]
@jyaunches  Refactored and pushed.

| Your point | Now |
|---|---|
| 477-line fork at `agents/hermes/plugin/googlechat_sandbox_adapter.py` | File deleted. The delta is the channel-owned `channels/googlechat/runtime/hermes-adapter.py`, 316 lines: REST pull, placeholder credentials, aiohttp transport, registry attach |
| Duplicated `connect()` at `:200` | `hermes-adapter.py:192` — a 9-line override that awaits `super().connect()` and starts the REST pull the bundled path skipped. Nothing copied |
| Four private methods rebound at `:392` | Gone. `grep -c MethodType` is 0 in both files; the delta is a subclass of `GoogleChatAdapter` overriding three methods |
| Global `platform_registry.register` wrapper at `:428` | Gone. `hermes-adapter.py:292-311` resolves the entry with `platform_registry.get()`, then `dataclasses.replace()` changes only `adapter_factory`, `check_fn` and `required_env`, then `register()`. Every other `PlatformEntry` field is carried through, and `get()` forces the deferred loader so registration order stops mattering |
| Conflicts with `src/lib/messaging/AGENTS.md:69-77` | Resolved. The module is a channel-owned runtime asset, its test sits beside it, and the Dockerfile copies it into the plugin directory at build time |
| "No Python or runtime test covers this layer" | Three test files, nine cases, all mutation-checked — below |

**Tests.**

- `channels/googlechat/runtime/hermes-adapter.test.ts` drives the real `_rest_pull` under `python3`, with doubles for aiohttp and the bundled adapter on `PYTHONPATH`. Asserts: every request carries only the credential placeholder; no URL beyond `:pull` and `:acknowledge` on the configured subscription; a nacked message sends no acknowledge; a rejected acknowledge and a failed acknowledge transport both leave the message for redelivery instead of ending the loop.
- `channels/googlechat/policy.test.ts` pins the egress preset: the two Pub/Sub operations, the Chat rules, the reachable host set, and no Pub/Sub egress for OpenClaw.
- `onboard/messaging-bridge-provider.test.ts` covers the bridge: one minted token carries every declared scope, and only the profile whose agent matches the sandbox produces a bridge.
- Each assertion was replayed against the mutation it exists to catch — changed placeholder, `:acknowledge` swapped for `:modifyAckDeadline`, a rejected acknowledge breaking the loop, the transport error escaping its handler, `nack()` acknowledging anyway, the agent filter removed, the scope list truncated to its first entry, Pub/Sub widened back to `/v1/**`, a third host added, Chat writes opened beyond the spaces tree. Each fails its own test and only that one.

**Line count.**

```
1116 total  -334 tests  -27 vendored wheels  =  755 reviewed surface
953 -> 755 = -198 (-21%)
```

**Partly done.**

- *The separate global-plugin gate.* Not removed.
  - 66 lines at your review, 41 now (`agents/hermes/plugin/__init__.py:1436`), and it loads the channel-owned module instead of a plugin-local fork.
  - It stays because Hermes loads only `__init__.py` as the plugin entry, so something has to load the sibling module.
  - Removing it outright means shipping the override as its own Hermes plugin directory. I have not verified plugins are individually enablable in the pinned release, so I am not offering that as the plan yet.
- *"If the pinned Hermes release lacks a stable seam, land that seam."* Split answer.
  - Attachment: the seam exists and this now uses it, which is why the registry wrapper is gone.
  - Transport and credentials: there is none — `PlatformEntry` carries no field for either — so the override subclasses `_validate_config`, `_load_sa_credentials` and `_new_authed_http`.
  - It patches no Hermes source, and `image-build-probes.py googlechat-override-seams` fails the image build if any of the five bound definitions moves, rather than silently restoring the stock gRPC and service-account adapter. `agents/hermes/` already carries nine build-time patches and sixteen such probes against Hermes internals.
  - If "land that seam" means an upstream Hermes change, say so and I will open it. I read the line as a condition rather than a request, so I did not assume.

**Missing coverage,** flagged rather than left to be found.

- **`UV_SYSTEM_CERTS` / `SSL_CERT_FILE`** — nothing asserts the uv install receives them. Without them uv trusts only its bundled roots and fails `UnknownIssuer` behind a MITM proxy. The fix is general rather than Google Chat's, so happy to split it out.
- **The `register()` gate** — nothing covers the subscription env set versus unset. A regression there leaves Hermes on the stock gRPC adapter, which the REST-only policy blocks, so the channel goes quiet instead of failing loudly. Needs the spawned-interpreter shape the adapter test uses, because CI does not run `agents/hermes/plugin/test_*.py`. The advisor raised this too.
- **E2E for receiving over Pub/Sub REST pull** — `channels-add-remove` and `messaging-providers` only cover OpenClaw's inbound webhook. A live target needs a real GCP project, subscription and key, so it is recorded as a combinatorial gap; the path was validated by hand end to end on two boxes.

**Review by jyaunches — COMMENTED** [review]
## LOC Reduction / Codebase Simplicity Review

Resolved at `8c5e6c4c25b1d3ce02b97ec8f4fe2bc58edd3072`.

The Google Chat delta now lives in the channel-owned runtime asset, subclasses the bundled adapter, calls the bundled `connect()`, and replaces only the published `PlatformEntry` fields through the documented registry seam. The global registry wrapper, copied connect implementation, instance method rebinding, and registration-order dependency are gone.

The remaining private Hermes method dependencies are explicit, image-build-pinned, and the REST pull, credential placeholder, allowed operations, acknowledgement behavior, and redelivery behavior now have deterministic runtime coverage.

No remaining blocking LOC or source-of-truth finding. This is a scope-limited follow-up, not an approval or a correctness, security, or CI review.

**Review by jyaunches — DISMISSED** [review]
## LOC Reduction / Codebase Simplicity Review

### What remains resolved

The earlier adapter-fork finding remains resolved at `9cd2fcc598dd972a975bd9ec2dc5df5bf48f9531`. The copied `connect()` implementation, global registry wrapper, private-method rebinding, and separate plugin-local adapter remain removed.

### Why changes are requested

The latest PR commit adds 170 lines and deletes 30. Of that delta, `src/lib/messaging/channels/googlechat/runtime/hermes-adapter.py:171,194-229` and its test add 97 net lines for a second pull-task handle and same-instance reconnect path.

`_sandbox_pull_task` and the bundled `_supervisor_task` now hold the same task. `_stop_rest_pull` exists because the bundled `connect()` clears its handle. However, the method documentation states that Hermes v2026.7.20 creates a fresh adapter for each reconnect. The new lifecycle authority and reconnect scenario therefore protect no current caller while expanding the private Hermes delta that the earlier refactor reduced.

The same commit also repeats agent selection at `src/lib/actions/sandbox/policy-channel.ts:862-879` and `src/lib/onboard/messaging-prep.ts:146-166`. Both callers normalize the agent, consult a manifest registry, map an unset agent to OpenClaw, map an unsupported agent to no bridge, and then call `collectMessagingBridgeTokenDefs`. That collector already owns the bridge profiles and filters them by `profile.agent` at `src/lib/onboard/messaging-bridge-provider.ts:260-279`.

### Refactor direction

- Remove the same-instance reconnect task state and reconnect-only test unless a current Hermes call path requires two `connect()` calls on one adapter.
- If that lifecycle is required, keep one task handle and show the reachable caller that needs replacement behavior.
- Let `collectMessagingBridgeTokenDefs` own the unset-agent default and unmatched-agent result once. Pass the recorded agent from both callers and let the profile filter select or reject it.

### Expected result

Keep the prior channel-owned adapter design, remove roughly 100 lines of unreachable reconnect scaffolding, and use one bridge-agent selection authority instead of two caller-specific branches.

**Review by jyaunches — COMMENTED** [review]
## LOC Reduction / Codebase Simplicity Review

### Resolution

The prior blocker is resolved at `44eab71c61ef882661bb2b875a578316468d5b64`.

- The latest PR commit removes `_sandbox_pull_task`, `_stop_rest_pull()`, and the same-instance reconnect scenario. The Hermes adapter again uses the bundled `_supervisor_task` as its only pull-task handle.
- `messagingBridgeProfilesForAgent()` now owns the unset-agent default and unsupported-agent result. Both token creation and provider reuse use that profile selection instead of repeating caller-specific branches.
- The follow-up adds 64 lines and deletes 150, for a net reduction of 86 lines.

### Result

The earlier channel-owned adapter refactor remains intact. The new reconnect authority and duplicate agent-selection branches are gone, with no replacement LOC or source-of-truth blocker in this review scope.

This resolves only the LOC reduction and codebase simplicity finding. It is not an approval or a correctness, security, or CI review.

File: src/lib/messaging/channels/googlechat/runtime/hermes-adapter.test.ts [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        "reconnect": [(200, {}, None), (200, delivery, None)] + [(200, {}, None)] * 8,
+    }[SCENARIO]
+)
+
+async def _until(predicate, limit=500):
+    for _ in range(limit):
+        if predicate():
+            return
+        await asyncio.sleep(0)
+    raise AssertionError("condition never held")
```
[the remark this answers is not in the document]
Comment (hunglp6d) [reply src/lib/messaging/channels/googlechat/runtime/hermes-adapter.test.ts]: @coderabbitai
Resolved by removal in `44eab71c`. The reconnect test this helper served is gone: in the pinned Hermes `v2026.7.20` both reconnect paths build a fresh adapter — `_create_adapter` at `gateway/run.py:8617` and `:9526`, before `connect(is_reconnect=True)` at `:8638` and `:9538` — so the same-instance guard protected no caller, and the maintainer review asked for it to go. The wall-clock point was correct while the test existed.

File: src/lib/messaging/channels/googlechat/runtime/hermes-adapter.test.ts [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    // intervening disconnect() would leave two consumers on one subscription and
+    // every message handled twice. The pinned gateway rebuilds the adapter per
+    // reconnect, so this pins the invariant rather than a reachable path.
+    const report = runScenario("reconnect");
+
+    expect(report.firstCancelled).toBe(true);
+    expect(report.replaced).toBe(true);
+    // The bundled connect() clears its own handle on this branch, so the
+    // replacement has to be rebound for the bundled disconnect() to cancel it.
+    expect(report.boundToBundledHandle).toBe(true);
```
[the remark this answers is not in the document]
Comment (hunglp6d) [reply src/lib/messaging/channels/googlechat/runtime/hermes-adapter.test.ts]: @coderabbitai
Removed in `44eab71c` together with the reconnect test and its doubles, so the private-shape assertions went with them. Recording that the point was right: before the removal I replaced `boundToBundledHandle` with an observable check — `disconnect()` must end a pull parked in a request that never returns — and only that version failed when the replacement task was not bound to the bundled handle. The field-identity assertion had let that mutation through.

File: src/lib/onboard/messaging-prep.ts [hunk 3]
```diff
@@ -138,16 +139,27 @@ export function prepareCreateSandboxMessaging(
...
+        sandboxName: input.sandboxName,
+        agent: bridgeAgent,
+        getCredential: input.getCredential,
+        env: input.env,
+        normalizeCredentialValue: input.normalizeCredentialValue,
+        enabledChannels: input.enabledChannels,
+        disabledChannelNames,
+      }),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (hunglp6d) [reply src/lib/onboard/messaging-prep.ts]: @coderabbitai
Fixed in `44eab71c`, though not with a separate gate. `messagingBridgeProfilesForAgent` is now the single selector: an unset agent is OpenClaw, an agent no profile declares selects nothing, and both the mint and the reuse pass read the same filtered profiles. The regression test asserts that `reusableMessagingProviders` and `reusableMessagingChannels` both exclude the Google Chat bridge for an unsupported agent.

**Review by prekshivyas — DISMISSED** [review]
Verdict: request changes on `44eab71c61ef882661bb2b875a578316468d5b64`.

The keyless transport is thoughtfully constrained: the service-account private key remains gateway-side, the sandbox receives only an OpenShell placeholder, the outbound hosts are fixed, Pub/Sub is limited to `:pull` and `:acknowledge`, and the added packages are pinned. Two blockers remain:

1. **The product scope is not accepted for Hermes Google Chat.** The linked #5492 is closed and its support table records Google Chat as OpenClaw `yes`, Hermes `no`, with NemoClaw Google Chat already completed by #7317. It does not accept this new Hermes runtime/design, and this PR has no closing issue. The repository Product Scope Gate requires an accepted issue or design decision naming the supported channel and agent runtime before implementation or approval. Please get explicit maintainer acceptance covering ownership, lifecycle, compatibility, security, and trusted validation before this is treated as canonical NemoClaw behavior.

2. **Bridge reuse is not bound to the selected agent profile.** `prepareCreateSandboxMessaging` filters profiles for the current agent, but its secretless resume path still accepts any existing `<sandbox>-googlechat-bridge` based only on name/existence. If a sandbox changes OpenClaw → Hermes (or the reverse), the durable provider can retain the previous agent's profile. In particular, the OpenClaw profile lacks Hermes's Pub/Sub scope, so resume can silently attach a credential that cannot pull events. Reuse must verify the existing provider's selected `profileId` and credential key, or provider identity must become agent-specific with stale-provider cleanup. Add a public-boundary agent-change/resume test with the source secret absent.

The unresolved CodeRabbit thread at `messaging-prep.ts:195` identifies the same second boundary. I left it unresolved. The current head also has failing CLI, managed-runtime, and advisor checks; those need a clean rerun after code changes. Once product scope is accepted, the linked design's lifecycle E2E and user documentation acceptance criteria also need to be completed.

Security rubric summary:

| Category | Status | Evidence |
|---|---|---|
| Injection / code execution | PASS | Fixed HTTPS hosts and structured request construction; policy fails closed outside allowed paths. |
| Secrets / data leakage | PASS | Raw service-account material is passed to gateway refresh by secret env and not rendered into the sandbox. |
| Authentication / authorization | FAIL | Durable provider reuse is not bound to the current agent/profile identity. |
| Cryptography | PASS | OAuth signing and token refresh remain at the gateway boundary. |
| Network / SSRF | PASS | Egress is restricted to Chat plus Pub/Sub pull/ack endpoints and exact Python executables. |
| Supply chain | PASS | Added wheels and agent package versions are pinned and checksum/integrity gated. |
| Configuration safety | FAIL | Agent changes can reuse stale bridge configuration with incompatible scopes. |
| Testing / verification | FAIL | No public-boundary agent-change resume regression; required lifecycle E2E/docs are not present. |
| Repository governance | FAIL | No accepted design currently authorizes Google Chat as a supported Hermes surface. |

I did not push an implementation because the repository's Product Scope Gate explicitly says to stop and request maintainer direction when this decision is missing.

**Comment by apurvvkumaria** [comment]
Merge-train blocker: **accepted product scope is required**

This PR creates a supported Google Chat integration for Hermes but does not link an accepted issue or design decision. The change adds 1,515 lines and removes 113 across 32 files, for a net increase of 1,402 lines. It also has four unresolved current review threads.

The repository product-scope gate requires a maintainer decision that defines ownership, lifecycle, compatibility, security, and validation expectations before this can become canonical NemoClaw behavior. CI repair alone cannot supply that decision.

Deferred for human direction. To resume the merge train, please link the accepted issue or design decision and confirm the intended release target. The current review findings and required checks must then pass on the resulting branch revision. No merge or close action is appropriate without that decision.

File: src/lib/messaging/channels/googlechat/runtime/hermes-adapter.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                    timeout=aiohttp.ClientTimeout(total=60),
+                ) as resp:
+                    content = await resp.read()
+                    info = {"status": resp.status}
+                    for key, value in resp.headers.items():
+                        info[key.lower()] = value
+                    return _gc.httplib2.Response(info), content
+
+        # Chat calls run under ``asyncio.to_thread``, so a fresh loop is safe here.
+        return asyncio.run(_run())
```
[the remark this answers is not in the document]
Comment (hunglp6d) [reply src/lib/messaging/channels/googlechat/runtime/hermes-adapter.py:145]: @coderabbitai
Checked against the release the image builds — Hermes `v2026.7.20`, the same tree `image-build-probes.py googlechat-override-seams` pins — and the premise does not hold: the bot-ID lookup does not run on the event-loop thread.

`_resolve_bot_user_id` (`plugins/platforms/google_chat/adapter.py:943`) dispatches the whole call through a worker:

```python
members = await asyncio.to_thread(          # :962
    lambda s=space: self._chat_api.spaces()
    .members()
    .list(parent=s, pageSize=50)
    .execute(http=self._new_authed_http())  # :966
)
```

That holds for every call site of `_new_authed_http()` in the release, not just this one:

| upstream | dispatch |
|---|---|
| `:966` `members.list` | `await asyncio.to_thread(lambda …)` |
| `:2318` `messages.delete` | `_do_delete` → `await asyncio.to_thread(_do_delete)` |
| `:2356` `messages.patch` | `_do_patch` → `await asyncio.to_thread(_do_patch)` |
| `:2605` `messages.create` | `_call_with_retry` → `:2549 return await asyncio.to_thread(sync_fn)` |
| `:3310` `spaces.get` | `await asyncio.to_thread(lambda …)` |

So `_GcAiohttpTransport.request()` always runs on a worker thread with no running loop, which is the precondition recorded at `hermes-adapter.py:144`, and `asyncio.run()` is correct there.

Consistent with that, the channel was validated end to end on a live sandbox: replies travel through this same transport, so a `RuntimeError` on it would leave the channel unable to answer at all. No change made.

**Comment by apurvvkumaria** [comment]
Update after the latest branch push: the product-scope blocker is unchanged. The PR still links no accepted issue or design decision, and the diff has grown to 1,568 additions and 113 deletions across 32 files, for a net increase of 1,455 lines.

Technical review and CI remediation remain deferred until a maintainer records the supported Google Chat ownership, lifecycle, compatibility, security, and validation contract. This update does not authorize merge or close action.

**Comment by hunglp6d** [comment]
@apurvvkumaria 
The accepted issue is #5492, now linked in the body with the scope reasoning and the release target.

That issue models support per agent and requires a channel to work for each runtime whose upstream supports it. Hermes `v2026.7.20` ships `plugins/platforms/google_chat/adapter.py`, so this is the Hermes half of an already accepted design, not a new supported surface. #7317 shipped the OpenClaw half and the issue was closed with only that half done; it is reopened and its table now records Google Chat as `yes` across all three columns against this PR. Ownership, lifecycle, security and validation are the ones it already defines. Release target: next patch release.

The figures read one revision behind: head `3668e10e9` is 1,573 additions and 113 deletions across 34 files, with no unresolved review thread.

If anything beyond #5492 must be recorded, name it and I will add it.

**Comment by hunglp6d** [comment]
@prekshivyas  Both blockers are addressed. Head is `3668e10e9`; your review was at `44eab71c6`.

**Product scope.** #5492 is the accepted issue: it models support per agent and requires a channel to work for each runtime whose upstream supports it, and Hermes `v2026.7.20` ships `plugins/platforms/google_chat/adapter.py`. #7317 delivered the OpenClaw half and the issue was closed with only that half done, which is why its table read Hermes `no`. It is reopened, the row now reads `yes` across all three columns against this PR, and the body carries the reasoning and the release target.

**Bridge reuse.** Fixed in `62d6e5dd9f`, one commit after the revision you reviewed. `messagingBridgeProfilesForAgent` is the single authority for which profiles an agent may use, the secretless resume path validates a candidate with `providerMatchesGatewayCredential(name, profileId, credentialKey)` rather than name and existence, preflight fails the run before onboard can delete and recreate the sandbox, and destroy derives the bridge suffixes from the profiles.

`messaging-prep.test.ts` covers reuse for the matching agent, refusal across agents, and refusal without an exact gateway binding. `test/onboard-pre-destructive-intent.test.ts` adds the public-boundary case you asked for: an agent change with the source secret absent, driven through a real onboard child process, asserted to exit before any mutating command. The CodeRabbit thread at `messaging-prep.ts:195` is resolved, and 27 checks are green with one managed-image build still running.

Live validation is recorded in the body, and a Google Chat E2E target needs a real GCP project, subscription and key.

**Comment by apurvvkumaria** [comment]
Product-scope review after the #5492 reference

I reviewed #5492 and #7317. #5492 is a design proposal with broad per-channel requirements, but it has no maintainer acceptance recorded for the Hermes Google Chat architecture. Its Hermes-specific reopening comment says the issue will track maintainer approval of that scope. Updating the channel table to “in progress” does not supply that approval.

#7317 explicitly scoped the shipped Google Chat behavior to OpenClaw and described Hermes as a later follow-up with a different inbound and credential model. This PR introduces that different supported surface: Pub/Sub REST pull, a gateway-minted multi-scope token, a Hermes runtime override, added image dependencies, new lifecycle behavior, and a live-validation gap that needs real GCP resources.

The product-scope gate therefore still needs a maintainer decision that accepts the Hermes design and names its ownership, compatibility, lifecycle, security, and release-validation expectations. The current +1,595/-130 change across 34 files is also flagged as large. Technical and CI work remains deferred because those choices can change the implementation and its required evidence. No merge or close action is appropriate until that human decision is recorded.

**Comment by sandl99** [comment]
Product scope clarification: Product Management approved Google Chat as a NemoClaw enterprise messaging channel for all supported agent runtimes, including Hermes. #7317 delivered OpenClaw first because OpenClaw and Hermes use different Google Chat protocols. This PR implements the Hermes portion of that approved scope. The decision is recorded in #5492: https://github.com/NVIDIA/NemoClaw/issues/5492#issuecomment-5355360216. The existing #5492 acceptance criteria and repository review gates still apply.

**Review by cv — APPROVED** [review]
Issue #5492 now records Product Management approval for first-class Google Chat across supported agent runtimes, including Hermes. The implementation keeps service-account material gateway-side, restricts egress to Chat and Pub/Sub operations, binds durable reuse to the selected agent profile and credential key, and records managed-Hermes live validation.
